### PR TITLE
ARROW-10019: [Rust] Add substring kernel

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR adds a new function to kernels called `substring` that yields a substring of a StringArray starting at a given index, and with a given optional length.

`fn substring(array: &Array, start: i32, length: &Option<u32>) -> Result<ArrayRef>`

This operation is common in strings, and it is useful for string-based transformations.

The API is inspired by Spark's API.

### CODE REVIEW

File: rust/arrow/src/compute/kernels/substring.rs
Comment (jhorstmann) [inline rust/arrow/src/compute/kernels/substring.rs:43]: I think we can estimate the capacity as `length * array.len()` if the length is specified and `values.len() - (start*array.len())` otherwise.

Hm, actually depends on the semantics of the `length` parameter, is it supposed to be a maximum length or should it be an error if `start+length` goes out of bounds of one entry?

Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/substring.rs:43]: I think that this all depends on what assumptions we make about a typical query. Specifically:

* what is the average start
* what is the average length
* what is the average string size

from these 3, we could compute an initial estimation. For example: is this going to be used to treat texts, on which "string size" can go to Mbs, or is this going to be used on 10-12 character arrays?

I considered this discussion beyond the scope of this PR and just went with the least memory-intensive/highest CPU-intensive approach, just because 100% memory usage usually has more nefarious implications than 100% CPU.

The semantics is so that it does not error: it just goes to the maximum/minimum available for each entry. The rational here is that otherwise the user would have to perform some calculation on the side to ensure no errors at runtime, which are never fun. :)

File: rust/arrow/src/compute/kernels/substring.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/substring.rs:88]: We should also support `DataType::LargeUtf8`. We're likely to get a scenario where Arrow users who deal with large data would want to use the `LargeList` and `LargeString` arrays by default (something I would also do), so it's ideal to support both string arrays from the onset going forward.

File: rust/arrow/src/compute/kernels/substring.rs
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/substring.rs:30]: @jorgecarleitao it might be possible to write this as a generic after the refactoring that you did for strings and lists. Do you want to try that? If so, I'll keep this PR open

**Review by nevi-me — APPROVED** [review]
LGTM, have a question which I'll wait for a response for before merging this.

File: rust/arrow/src/compute/kernels/substring.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/substring.rs]: done :)

**Comment by nevi-me** [comment]
Just checked out the PR and ran a clippy check. The PR and a few others before have introduced a few warnings and errors.
I find clearing lints helps me learn interesting shortcuts along the line, @jorgecarleitao do you mind checking the lints and clearing them (even those beyond the scope of this PR if you'd like)?

The most interesting for me was learning that to clone an optional reference (`let some_struct_ref = Option<&MyStruct>`), you can use `let some_struct: Option<MyStruct> = some_struct_ref.cloned()` instead of `some_struct_ref.map(|s| s.clone())`.

**Comment by jorgecarleitao** [comment]
@nevi-me , no problem, clippy is important. I fixed the ones that `cargo clippy` showed as a separate (last) commit in this PR.

**Pull request closed without merging**
